### PR TITLE
SITL: Make AP_* libraries Singletons.

### PR DESCRIPTION
**Pull request by mayanez**
Additionally update `AP_Baro/examples/BARO_generic` to support new usage.

### CODE REVIEW

File: libraries/AP_HAL_SITL/SITL_State.cpp [hunk 1]
```diff
@@ -75,13 +75,13 @@ void SITL_State::_sitl_setup(const char *home_str)
...
     fprintf(stdout, "Starting SITL input\n");
 
+    _sitl = SITL::SITL::instance();
     // find the barometer object if it exists
-    _sitl = (SITL::SITL *)AP_Param::find_object("SIM_");
-    _barometer = (AP_Baro *)AP_Param::find_object("GND_");
-    _ins = (AP_InertialSensor *)AP_Param::find_object("INS_");
-    _compass = (Compass *)AP_Param::find_object("COMPASS_");
+    _barometer = AP_Baro::instance();
+    _ins = AP_InertialSensor::get_instance();
```
Comment (mayanez) [inline libraries/AP_HAL_SITL/SITL_State.cpp]: AP_InertialSensor was already a singleton hence the different function naming. I wasn't sure whether I should update it accordingly to match the rest. Let me know and I'll quickly update it.

Comment (mayanez) [inline libraries/AP_HAL_SITL/SITL_State.cpp]: The 4.9.3 compiler was giving errors on a namespace collision with `instance` where as 5.4 was not. I've changed them to match `get_instance` like `AP_InertialSensor`.

**Comment by OXINARF** [comment]
Thanks for this @mayanez. The first commit (together with the last) needs to be separated in multiple commits, one for each library. I think @tridge already mentioned it to you but we have scripts that help do this: https://github.com/ArduPilot/ardupilot/tree/master/Tools/gittools

Before making more changes we should wait for @tridge or @lucasdemarchi's opinion (and possibly others), but I think that the `get_instance` should always return an instance, just like the AP_InertialSensor already does - that is the principle of a singleton.

**Comment by mayanez** [comment]
Ah wasn't aware there was a tool for that. I'll make sure to use it from now on.

Yes, I think there are a few ways to create singleton's with `AP_InertialSensor` being the lazy instantiation way. Definitely people with some more C++ background should chime in. I know there are issues with thread safety etc. `AP_Terrain` is probably the one that is a bit tricky since it takes in references which makes initialization a bit of a pain.

**Comment by mayanez** [comment]
statically defining the instances was causing the flash on PX4 to overflow. I've changed it to how `AP_Inertialsensor` with `new`. I wonder if we should consider `#ifdef` `get_instance()`. Any thoughts.

**Comment by lucasdemarchi** [comment]
> statically defining the instances was causing the flash on PX4 to overflow. I've changed it to how AP_Inertialsensor with new. I wonder if we should consider #ifdef get_instance(). Any thoughts.

This should actually not be used with new(), unless there's also enough memory from heap and we really want to do this.  The point in being static is exactly that then it will live in flash for PX4-based boards.  It passes the 1MB limit because we are very close to that limit and just adding some pointers is sufficient to reach that limit.

Instead of allocating from heap, the solution would be to 1) reduce flash size from other parts and 2) reduce flash size needed by this code. For (1), #6262 is probably sufficient since it's giving us 1K of space.... but we can check other options, too.

For (2):

There are various flavors of singleton implementations. Giving our requirements, I would separate the initialization from the get_instance() calls. This allows the compiler to avoid the inline code to check for initialization that needs to be added everywhere. I did exactly that on mavlink-router recently for different reasons (see https://github.com/01org/mavlink-router/commit/99d677d83ab8fd499ff5c5d783785421e8af02f6). If you move the calls back to avoid using new() I can add some commits to improve it.

**Comment by mayanez** [comment]
Hey @lucasdemarchi thanks for the comments.
Yes, I had created the instances statically as you did in that commit, but ran into the issue of overflowing flash memory. My suggestion on the `#ifdefs` was basically to avoid having `new` be called from PX4 assuming we dont need `get_instance()` there. I'll switch to the static implementation again that way you can let me know if we can make it fit under 1MB.

**Comment by lucasdemarchi** [comment]
The `get_instance()` method should not be defined only for sitl.... that would create lots of  headache of maintenance.  We can use the same approach for the other HALs since we can avoid most of the flash overhead from this current implementation.

**Comment by mayanez** [comment]
I changed it up a bit by instead having the parameter map point to the instances. I'm not sure why its not working on anything but Copter. Somehow the parameters are being treated differently in other vehicles. Any suggestions?

**Pull request closed without merging**